### PR TITLE
Add support methods for CR objects

### PR DESCRIPTION
**Pull request by jsmolar**
This PR adds dunder `__getitem__` and `__setitem__` methods for CR objects. Also Implements authorino-like and limitador-like objects for Kuadrant CR to enable similar manipulation with Limitador and sub-component Limitador from Kuadrant CR.
I had to remove `from testsuite.config import settings` from utils because of cyclical import.

### CODE REVIEW

File: testsuite/openshift/__init__.py [hunk 1]
```diff
@@ -44,6 +46,25 @@ def wait_until(self, test_function, timelimit=90):
             return False
 
 
+class CRObject(OpenShiftObject):
```
Comment (pehala) [inline testsuite/openshift/__init__.py:49]: How about naming it `CustomResource`?

File: testsuite/openshift/__init__.py [hunk 2]
```diff
@@ -44,6 +46,25 @@ def wait_until(self, test_function, timelimit=90):
...
+    """Custom APIObjects that implements methods that improves manipulation with CR objects"""
+
+    def safe_apply(self):
+        """Reimplementation of modify_and_apply from OpenshiftObject"""
+        result, status = self.modify_and_apply(lambda _: True, retries=2)
+        assert status
```
Comment (pehala) [inline testsuite/openshift/__init__.py:55]: I would add message to the assert to better indicate what went wrong

File: testsuite/openshift/__init__.py [hunk 3]
```diff
@@ -44,6 +46,25 @@ def wait_until(self, test_function, timelimit=90):
             return False
 
 
+class CRObject(OpenShiftObject):
+    """Custom APIObjects that implements methods that improves manipulation with CR objects"""
+
+    def safe_apply(self):
+        """Reimplementation of modify_and_apply from OpenshiftObject"""
```
Comment (pehala) [inline testsuite/openshift/__init__.py:53]: Better description

File: testsuite/openshift/kuadrant.py [hunk 4]
```diff
@@ -1,29 +1,87 @@
...
+            self.kuadrant_cr.model.spec[self.spec_name][name] = dataclasses.asdict(value)
+        else:
+            self.kuadrant_cr.model.spec[self.spec_name][name] = value
+
+    def wait_for_ready(self):
+        """Waits until Kuadrant CR reports ready status"""
+        self.kuadrant_cr.wait_for_ready()
+
+
+class _Authorino(_CRObject):
```
Comment (pehala) [inline testsuite/openshift/kuadrant.py:50]: No need to keep this private

Comment (pehala) [inline testsuite/openshift/kuadrant.py:50]: I would not introduce these subclasses now, we can add them later when they are actually needed

File: testsuite/openshift/kuadrant.py [hunk 5]
```diff
@@ -1,29 +1,87 @@
...
+class KuadrantCR(CRObject):
     """Represents Kuadrant CR objects"""
 
     LIMITADOR = "limitador-limitador"
 
-    @property
-    def limitador(self) -> dict:
-        """Returns spec.limitador from Kuadrant object"""
-        return self.model.spec.setdefault("limitador", {})
+    @cached_property
```
Comment (pehala) [inline testsuite/openshift/kuadrant.py:63]: Wont hurt but not sure if needed, there is not state saved in the class and it is inexpensive operation

File: testsuite/openshift/limitador.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class LimitadorCR(CRObject):
+    """Represents Limitador CR objects"""
+
+    @property
+    def deployment(self) -> Deployment:
+        """Returns Deployment object for this Limitador"""
+        with self.context:
+            return selector("deployment/limitador-limitador").object(cls=Deployment)
```
Comment (pehala) [inline testsuite/openshift/limitador.py:16]: Name should be dynamic

File: testsuite/tests/conftest.py [hunk 7]
```diff
@@ -226,6 +226,9 @@ def oidc_provider(keycloak) -> OIDCProvider:
 @pytest.fixture(scope="session")
 def blame(request):
     """Returns function that will add random identifier to the name"""
+    if "tester" in settings:
```
Comment (pehala) [inline testsuite/tests/conftest.py:229]: unrelated change, probably deserves a separate PR

File: testsuite/openshift/kuadrant.py [hunk 8]
```diff
@@ -1,29 +1,87 @@
...
-    def limitador(self, value: dict):
-        """Sets the value of spec.limitador"""
-        self.model.spec["limitador"] = value
+    @cached_property
+    def limitador(self) -> _Limitador:
+        """Returns spec.limitador from Kuadrant object"""
+        return _Limitador(self, "limitador")
 
     @property
     def limitador_deployment(self):
```
Comment (pehala) [inline testsuite/openshift/kuadrant.py:74]: This is not consistent with LimitadorCR

File: testsuite/tests/kuadrant/reconciliation/cr_configuration/test_reconcile_limitador.py [hunk 9]
```diff
@@ -18,31 +21,43 @@ def kuadrant_clean_up(request, kuadrant):
...
+    value = asdict(ContainerResources(requests_cpu="300m", requests_memory="40Mi"))
+
+    kuadrant.limitador["resourceRequirements"] = value
+    kuadrant.safe_apply()
+    kuadrant.wait_for_ready()
 
-    assert value == kuadrant.limitador[field]
-    assert str(value) in str(kuadrant.limitador_deployment.model.spec)
+    assert kuadrant.limitador["resourceRequirements"] == value
+    assert kuadrant.limitador_deployment.model.spec.template.spec.containers[0]["resources"] == value
```
Comment (pehala) [inline testsuite/tests/kuadrant/reconciliation/cr_configuration/test_reconcile_limitador.py:60]: I would use asdict here instead of above

File: testsuite/openshift/kuadrant.py [hunk 10]
```diff
@@ -1,29 +1,78 @@
 """Kuadrant CR object"""
 
-from openshift_client import selector
+import dataclasses
```
[the remark this answers is not in the document]
Comment (jsmolar) [reply testsuite/openshift/kuadrant.py:3]: for set item: `if dataclasses.is_dataclass(value):`

File: testsuite/openshift/kuadrant.py (hunk 8 above)
Comment (jsmolar) [reply testsuite/openshift/kuadrant.py:56]: I removed LimitadorCR as it is not used in this PR

File: testsuite/openshift/kuadrant.py [hunk 11]
```diff
@@ -1,29 +1,69 @@
...
+
+class KuadrantSection:
+    """
+    Base class for Kuadrant sub components:
+        Authorino - spec.authorino
+        Limitador - spec.limitador
+    """
+
+    def __init__(self, kuadrant_cr, spec_name):
+        super().__init__()
```
[the remark this answers is not in the document]
Comment (jsmolar) [reply testsuite/openshift/kuadrant.py:20]: We discussed this with Petr, that this is a good practice

**Review by averevki — APPROVED** [review]

**Review by pehala — APPROVED** [review]
